modules/images/image-loading-optimization/class-ilo-url-metrics-group-collection.php

```diff
@@ -0,0 +1,241 @@
+	 */
+	private $groups;
+
+	/**
+	 * Breakpoints in max widths.
+	 *
+	 * @var int[]
+	 * @phpstan-var int<1, max>[]
+	 */
+	private $breakpoints;
+
+	/**
+	 * Sample size for URL metrics for a given breakpoint.
+	 *
+	 * @var int
+	 * @phpstan-var int<1, max>
+	 */
+	private $sample_size;
+
+	/**
+	 * Freshness age (TTL) for a given URL metric.
+	 *
+	 * @var int
+	 * @phpstan-var int<0, max>
+	 */
+	private $freshness_ttl;
+
+	/**
+	 * Constructor.
+	 *
+	 * @throws InvalidArgumentException When an invalid argument is supplied.
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics   URL metrics.
+	 * @param int[]            $breakpoints   Breakpoints in max widths.
+	 * @param int              $sample_size   Sample size for the maximum number of viewports in a group between breakpoints.
+	 * @param int              $freshness_ttl Freshness age (TTL) for a given URL metric.
+	 */
+	public function __construct( array $url_metrics, array $breakpoints, int $sample_size, int $freshness_ttl ) {
+		// Set breakpoints.
+		sort( $breakpoints );
+		$breakpoints = array_values( array_unique( $breakpoints, SORT_NUMERIC ) );
+		foreach ( $breakpoints as $breakpoint ) {
+			if ( $breakpoint <= 0 ) {
+				throw new InvalidArgumentException(
+					esc_html(
+						sprintf(
+							/* translators: %d is the invalid breakpoint */
+							__(
+								'Each of the breakpoints must be greater than zero, but encountered: %d',
+								'performance-lab'
+							),
+							$breakpoint
+						)
+					)
+				);
+			}
+		}
+		$this->breakpoints = $breakpoints;
+
+		// Set sample size.
+		if ( $sample_size <= 0 ) {
+			throw new InvalidArgumentException(
+				esc_html(
+					sprintf(
+						/* translators: %d is the invalid sample size */
+						__( 'Sample size must be greater than zero, but provided: %d', 'performance-lab' ),
+						$sample_size
+					)
+				)
+			);
+		}
+		$this->sample_size = $sample_size;
+
+		// Set freshness TTL.
+		if ( $freshness_ttl < 0 ) {
+			throw new InvalidArgumentException(
+				esc_html(
+					sprintf(
+						/* translators: %d is the invalid sample size */
+						__( 'Freshness TTL must be at least zero, but provided: %d', 'performance-lab' ),
+						$freshness_ttl
+					)
+				)
+			);
+		}
+		$this->freshness_ttl = $freshness_ttl;
+
+		// Create groups and the URL metrics to them.
+		$this->groups = $this->create_groups();
+		foreach ( $url_metrics as $url_metric ) {
+			$this->add_url_metric( $url_metric );
+		}
+	}
+
+	/**
+	 * Create groups.
+	 *
+	 * @return ILO_URL_Metrics_Group[] Groups.
+	 */
+	private function create_groups(): array {
+		$groups    = array();
+		$min_width = 0;
+		foreach ( $this->breakpoints as $max_width ) {
+			$groups[]  = new ILO_URL_Metrics_Group( array(), $min_width, $max_width, $this->sample_size, $this->freshness_ttl );
+			$min_width = $max_width + 1;
+		}
+		$groups[] = new ILO_URL_Metrics_Group( array(), $min_width, PHP_INT_MAX, $this->sample_size, $this->freshness_ttl );
```

Comment:
I see. Just thinking out loud, would it be more straightforward to just add `PHP_INT_MAX` to `$breakpoints` in the constructor instead?

Reply:
I don't think so per https://github.com/WordPress/performance/pull/1010#discussion_r1515634477.